ejb3/src/main/java/org/jboss/as/ejb3/remote/AssociationImpl.java

```diff
@@ -73,23 +73,59 @@
 import java.util.Map;
 import java.util.Set;
 import java.util.concurrent.CancellationException;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.Executor;
 import java.util.concurrent.Future;
 import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
 
 /**
  * @author <a href="mailto:tadamski@redhat.com">Tomasz Adamski</a>
  */
-final class AssociationImpl implements Association {
+final class AssociationImpl implements Association, Registry.Listener<String, List<ClientMapping>>, AutoCloseable {
```

Comment:
"I'd prefer to stick with composition for the listener instead of inheritance in this particular case."
OK.

"I don't think that the fact that events are queued sequentially and the fact that they are per-connection are logically connected in any way (each connection independently has a sequential view, and each connection is separately asynchronous in terms of when it receives/processes events), thus from that perspective it doesn't seem to matter whether there is a listener per connection or a single listener which is delegated to many connections."
Each Registry listener is allocated a blocking queue and single thread for processing topology change events, as Registry.Listeners are contractually non-interfering.  Given that the number of connections can be quite large, it doesn't make sense to register a Registry.Listener per connection.

"Also it's worth noting that there is exactly one consumer of the Registry API."
On the contrary, each Undertow server uses a Registry to lookup routing identifiers for each cluster member.  Additionally, as part of the public clustering API, the Registry API is exposed to deployments.  The singleton domain controller will likely be another consumer.